### PR TITLE
fix: generated aliases for struct field access no longer require quoting

### PR DESCRIPTION
**Pull request by big-andy-coates**
### Description 

fixes: #4911

With this change there will no longer be the need to quote the column names generated for struct fields, unless the struct field name itself requires quoting. For example:

```sql
SELECT someStruct->mapOfStructsField['someKey']->arrayOfStructsFields[10]->someField FROM FOO;
```

Previously the above would generate a column name that would require quoting to be used: `SOMESTRUCT__MAPOFSTRUCTSFIELD['someKey']__ARRAYOFSTRUCTSFIELD[10]__SOMEFIELD`. It will now generate the name `SOMEFIELD`.

BREAKING CHANGE: This commit changes the system generated column name for any columns in projections that are struct field dereferences. Previously, the full path was used when generating the name, now only the final field name is used. For example, `SELECT someStruct->someField, ...` would previously of generated a column name of `SOMESTRUCT__SOMEFIELD`, and will now generate a name of `SOMEFIELD`. Generated column names may have a numeric appended to the end to ensure uniqueness, for example `SOMEFIELD_2`.

Note: it is recommended that you do not rely on system generated column names for production systems as there naming may change between releases. Providing an explict alias ensures consistent naming across releases. For example, `SELECT someStruct->someField AS someField`.

Backwards compatibility: existing running queries will not be affected by this change: they will continue to run with the same column names. Any statements executed after the upgrade will use the new names where no explicit alias is provided. Add explicit aliases to your statements if you require the old names, for example: `SELECT someStruct->someField AS SOMESTRUCT__SOMEFIELD, ...`

### Reviewing notes:

Commits:
 1. Prod code changes.
 1. Test code changes.
 1. Historical plans changes: the only change is the name of the generated column names.

For the prod changes I'd suggest take a look at the following in this rough order:
  1. I've added a new `ColumnAliasGenerate` interface.
  1. I've updated `ColumnNames` to return the above from `columnAliasGenerator`. Internally, there's a new `StructFieldAliasGenerator` class which is responsible for ensuring, (to a point), that generated aliases for struct field access are unique.  Please take a look at the unit test for this class and see if you can think of any test cases I've missed!  This is the bit of the code that is most likely to have unforeseen edge cases!!!
   1. I've wired in the new `ColumnAliasGenerate` where ever the old `ColumnNames.columnAliasGenerator` or `ColumnNames. generatedStructFieldColumnName` was being used: the new interface does both.

### Testing done 

Usual.

### Reviewer checklist
- [ ] Ensure docs are updated if necessary. (eg. if a user visible feature is being added or changed).
- [ ] Ensure relevant issues are linked (description should include text like "Fixes #<issue number>")

### CODE REVIEW

File: ksqldb-common/src/test/java/io/confluent/ksql/name/ColumnNamesTest.java [hunk 1]
```diff
@@ -1,90 +0,0 @@
...
-import io.confluent.ksql.schema.ksql.types.SqlTypes;
-import io.confluent.ksql.util.KsqlException;
-import java.util.List;
-import java.util.function.Supplier;
-import java.util.stream.Collectors;
-import java.util.stream.IntStream;
-import java.util.stream.Stream;
-import org.junit.Test;
-
-public class ColumnNamesTest {
```
Comment (big-andy-coates) [inline ksqldb-common/src/test/java/io/confluent/ksql/name/ColumnNamesTest.java:35]: This hasn't been deleted. Git just hasn't picked up the file move...

File: ksqldb-execution/src/main/java/io/confluent/ksql/name/ColumnNames.java [hunk 2]
```diff
@@ -17,28 +17,30 @@
...
 public final class ColumnNames {
 
   private static final String AGGREGATE_COLUMN_PREFIX = "KSQL_AGG_VARIABLE_";
-  private static final String GENERATED_ALIAS_PREFIX = "KSQL_COL_";
+  private static final String GENERATED_ALIAS_PREFIX = "KSQL_COL";
   private static final String SYNTHESISED_COLUMN_PREFIX = "KSQL_SYNTH_";
 
-  private static final Pattern GENERATED_ALIAS_PATTERN = Pattern
-      .compile(GENERATED_ALIAS_PREFIX + "(\\d+)");
+  private static final Pattern NUMBERED_COLUMN_PATTERN = Pattern.compile("(.*?)?(?:_(\\d+))?");
```
Comment (agavra) [inline ksqldb-execution/src/main/java/io/confluent/ksql/name/ColumnNames.java:43]: can we have a comment explaining the regex? also instead of relying on `?:` and indexed groups we should take advantages of named groups (https://docs.oracle.com/javase/7/docs/api/java/util/regex/Matcher.html#group%28java.lang.String%29)

Comment (big-andy-coates) [reply ksqldb-execution/src/main/java/io/confluent/ksql/name/ColumnNames.java:43]: Your wish....

**Comment by SunQK0515** [comment]
Hi, if i use this create,"CREATE STREAM TEST_05180004 AS SELECT AFTER -> AAA_2 FROM xxx ;"
But the result column is "AAA". It should be "AAA_2",do you know why,and how should i do can get except result?
